### ISSUE TITLE
Rendering glitch with Lens Holder

### ISSUE THREAD

**Issue by Da-Technomancer** [Da-Technomancer/Crossroads#23]
If I fire a beam at a lens holder containing a gem that filters magic, ex. ruby, and the beam contains that element (energy) as well as other elements, then the beam renders pure red for the entire length, instead of being the color mix (ex yellow) until the lens, then being pure red past that point. The cause of this problem is that in the standard emit method the beam gets assigned a new MagicUnit when it passes the lens, and doesn't use the MagicUnit color until after it finds the beam's end point, therefore using the new color for the entire beam. Problematically, if the color is found at the beginning instead then the whole beam will be the original color instead, which is arguably worse. With the current BeamRender system, the entire beam has to be the same color. The only two solutions I can think of are too either change the BeamRender system to take a list of triples for segments of the beam, or to make the lens absorb the old beam and fire a new one. The second option is simpler but the first option would be preferred.

**Comment by Da-Technomancer** [comment]
After looking into it further, it seems that changing BeamRender to work that way is not trivial. The major problem is scalability: All the data for one beam has to be sent in a single packet for reliability. The current method of using an int can only fit one beam segment, and using a long limits this to two beam segments. Using a String is the only way to have high scalability, and that would require complicated changes to how the packets are created and processed. It's not worth doing it by this method.

Changing the lens holder to fire a new beam is possible, infact it pretty simple. However, that would mean decreasing the reliability of the lens holder for careful timing. Considering the lens is supposed to be reliable, this is not a great option.

In summary, this bug can stay. It's not a huge problem, and maybe it will get fixed at some major re-write of the magic system at some point, but until then I'll just work it into the lore somehow.

**Comment by Da-Technomancer** [comment]
This is turning into an entire rewrite of the magic system, but I think the improvements will be worth it.
